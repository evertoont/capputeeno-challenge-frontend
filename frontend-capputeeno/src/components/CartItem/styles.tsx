import styled from "styled-components";

export const CartItemContainer = styled.div`
  display: flex;
  align-items: center;
  border-radius: 8px;
  background-color: white;
  /* width: 736px; */
  height: 210px;
`;

export const ProductImage = styled.img`
  max-height: 100%;
  width: 256px;
  border-radius: 8px 0 0 8px;
`;

export const ProductInfoContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 16px 31px;
  height: 100%;
  width: 100%;
`;

export const ProductInfoHeader = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const ProductInfoTitle = styled.h4`
  color: ${({ theme }) => theme.colors.secondTextDark};
  font-size: 20px;
  font-weight: 300;
  line-height: 150%;
`;

export const ProductDeleteButton = styled.button`
  width: 24px;
  height: 24px;
  background-color: transparent;
  border: none;
  cursor: pointer;
`;

export const ProductInfoDescription = styled.p`
  color: ${({ theme }) => theme.colors.secondTextDark};
  font-weight: 400;
  font-size: 12px;
  max-height: 50%;
  overflow: hidden;
  text-overflow: ellipsis;
  margin-top: 12px;
`;

export const ProductInfoFooter = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const ProductInfoPrice = styled.span`
  color: ${({ theme }) => theme.colors.shapesDark};
  font-size: 16px;
  font-weight: 600;
  line-height: 150%;
`;

export const SelectQuantity = styled.select`
  padding: 8px;
  max-width: 65px;
  width: 100%;
  border: 1.5px solid ${({ theme }) => theme.colors.borderColor};
  border-radius: 8px;
  background-color: ${({ theme }) => theme.colors.secondary};
  color: ${({ theme }) => theme.colors.textDark};
  font-weight: 400;
  font-size: 16px;
  outline: none;
`;
