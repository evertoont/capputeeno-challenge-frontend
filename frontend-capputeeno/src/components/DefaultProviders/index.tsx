"use client";

import { theme } from "@/app/theme";
import { FilterContextProvider } from "@/context/filterContext";
import { ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { ThemeProvider } from "styled-components";

interface DefaultProvidersProps {
  children: ReactNode;
}
export function DefaultProviders(props: DefaultProvidersProps) {
  const client = new QueryClient();

  return (
    <QueryClientProvider client={client}>
      <FilterContextProvider>
        <ThemeProvider theme={theme}>{props.children}</ThemeProvider>
      </FilterContextProvider>
    </QueryClientProvider>
  );
}
