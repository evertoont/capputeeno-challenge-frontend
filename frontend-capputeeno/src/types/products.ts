export interface Product {
  name: string;
  price_in_cents: number;
  id: string;
  image_url: string;
  description?: string;
  category?: string;
}

export interface ProductsFetchResponse {
  data: {
    allProducts: Product[];
    _allProductsMeta: {
      count: number;
    };
  };
}

export interface ProductFetchResponse {
  data: {
    Product: Product;
  };
}

export interface ProductInCart extends Product {
  quantity: number;
}
