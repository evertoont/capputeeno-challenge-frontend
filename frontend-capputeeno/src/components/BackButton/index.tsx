import { useRouter } from "next/navigation";
import { ButtonText, ContainerButton } from "./styles";
import { BackIcon } from "../icons/back";

interface BackButtonProps {
  navigateTo: string;
}
export function BackButton(props: BackButtonProps) {
  const { navigateTo } = props;
  const router = useRouter();

  const handleBackButton = () => {
    router.push(navigateTo);
  };

  return (
    <ContainerButton onClick={handleBackButton}>
      <BackIcon />
      <ButtonText>Voltar</ButtonText>
    </ContainerButton>
  );
}
