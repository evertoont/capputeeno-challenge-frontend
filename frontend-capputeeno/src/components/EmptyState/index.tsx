import { Empty } from "../icons/empty";
import { EmptyContainer, EmptyMessage } from "./styles";

interface EmptyStateProps {
  children?: React.ReactNode;
}

export function EmptyState(props: EmptyStateProps) {
  const { children } = props;

  return (
    <EmptyContainer data-testid="EmptyState">
      <EmptyMessage>{children}</EmptyMessage>
      <Empty />
    </EmptyContainer>
  );
}
