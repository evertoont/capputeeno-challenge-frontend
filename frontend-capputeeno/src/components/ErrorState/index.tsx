import { Error } from "../icons/error";
import { ErrorContainer, ErrorMessage } from "./styles";

interface EmptyStateProps {
  children?: React.ReactNode;
}

export function ErrorState(props: EmptyStateProps) {
  const { children } = props;

  return (
    <ErrorContainer data-testid="ErrorState">
      <ErrorMessage>{children}</ErrorMessage>
      <Error />
    </ErrorContainer>
  );
}
