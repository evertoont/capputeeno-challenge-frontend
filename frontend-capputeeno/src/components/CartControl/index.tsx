"use client";

import { useRouter } from "next/navigation";
import { CartIcon } from "../icons/cart";
import { CartCount, CartContainer } from "./styles";
import { useReadLocalStorage } from "usehooks-ts";
import { ProductInCart } from "@/types/products";

export function CartControl() {
  const router = useRouter();
  const productCartList = useReadLocalStorage<ProductInCart[]>("product-cart");

  const existProductCart = productCartList && productCartList.length > 0;

  const handleClickCart = () => {
    router.push("/cart");
  };

  return (
    <CartContainer onClick={handleClickCart}>
      <CartIcon />
      {existProductCart && <CartCount>{productCartList?.length}</CartCount>}
    </CartContainer>
  );
}
