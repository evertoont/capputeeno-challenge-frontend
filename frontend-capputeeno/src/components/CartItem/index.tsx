import { ProductInCart } from "@/types/products";
import {
  CartItemContainer,
  ProductDeleteButton,
  ProductImage,
  ProductInfoContainer,
  ProductInfoDescription,
  ProductInfoFooter,
  ProductInfoHeader,
  ProductInfoPrice,
  ProductInfoTitle,
  SelectQuantity,
} from "./styles";
import { DeleteIcon } from "../icons/delete";
import { formatPrice } from "@/utils/produtcs";
import { useProductSettings } from "@/hooks/useProductSettings";

interface CartItemProps {
  productInfo: ProductInCart;
  onRemoveProduct: (productId: string) => void;
  onUpdateQuantity: (productId: string, quantity: number) => void;
}
export function CartItem(props: CartItemProps) {
  const { productInfo, onRemoveProduct, onUpdateQuantity } = props;
  const { maxQuantityItems } = useProductSettings();

  const handleUpdateQuantity = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    onUpdateQuantity(productInfo.id, Number(event.target.value));
  };

  return (
    <CartItemContainer>
      <ProductImage src={productInfo.image_url} />
      <ProductInfoContainer>
        <ProductInfoHeader>
          <ProductInfoTitle>{productInfo.name}</ProductInfoTitle>
          <ProductDeleteButton onClick={() => onRemoveProduct(productInfo.id)}>
            <DeleteIcon />
          </ProductDeleteButton>
        </ProductInfoHeader>

        <ProductInfoDescription>
          {productInfo.description}
        </ProductInfoDescription>

        <ProductInfoFooter>
          <SelectQuantity
            value={productInfo.quantity}
            onChange={handleUpdateQuantity}
          >
            {Array.from({ length: maxQuantityItems }, (_, index) => (
              <option key={index} value={index + 1}>
                {index + 1}
              </option>
            ))}
          </SelectQuantity>
          <ProductInfoPrice>
            {formatPrice(productInfo.price_in_cents)}
          </ProductInfoPrice>
        </ProductInfoFooter>
      </ProductInfoContainer>
    </CartItemContainer>
  );
}
