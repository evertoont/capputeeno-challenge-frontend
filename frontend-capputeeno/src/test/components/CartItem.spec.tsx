import { fireEvent, render } from "@testing-library/react";
import { productInCartMock } from "../mocks/products";
import { CartItem } from "@/components/CartItem";
import { DefaultProviders } from "@/components/DefaultProviders";

const mockUpdateQuantity = jest.fn();
const mockRemoveProduct = jest.fn();

jest.mock("@/utils/products", () => ({
  formatPrice: jest.fn().mockReturnValue("R$ 10,00"),
}));

jest.mock("@/hooks/useProductSettings", () => ({
  useProductSettings: jest.fn().mockReturnValue({
    maxQuantityItems: 10,
  }),
}));

describe("CartItem", () => {
  it("should render CartItem correctly", () => {
    const { getByTestId } = render(
      <DefaultProviders>
        <CartItem
          onRemoveProduct={mockRemoveProduct}
          onUpdateQuantity={mockUpdateQuantity}
          productInfo={productInCartMock}
        />
      </DefaultProviders>
    );

    expect(getByTestId("CartItem")).toMatchSnapshot();
  });

  it("should call onUpdateQuantity when select quantity", () => {
    const { getByRole, debug } = render(
      <DefaultProviders>
        <CartItem
          onRemoveProduct={mockRemoveProduct}
          onUpdateQuantity={mockUpdateQuantity}
          productInfo={productInCartMock}
        />
      </DefaultProviders>
    );

    const selectQuantity = getByRole("combobox");

    expect(selectQuantity).toBeInTheDocument();

    fireEvent.change(selectQuantity, { target: { value: "2" } });

    expect(mockUpdateQuantity).toHaveBeenCalledTimes(1);
  });

  it("should call onRemoveProduct when click on delete button", () => {
    const { getByRole } = render(
      <DefaultProviders>
        <CartItem
          onRemoveProduct={mockRemoveProduct}
          onUpdateQuantity={mockUpdateQuantity}
          productInfo={productInCartMock}
        />
      </DefaultProviders>
    );

    const deleteButton = getByRole("button");

    expect(deleteButton).toBeInTheDocument();

    fireEvent.click(deleteButton, "click");

    expect(mockRemoveProduct).toHaveBeenCalledTimes(1);
  });
});
