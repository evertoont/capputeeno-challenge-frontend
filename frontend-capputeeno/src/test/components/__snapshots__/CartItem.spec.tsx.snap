// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem should render CartItem correctly 1`] = `
<div
  class="sc-bcPKhP goIgTE"
  data-testid="CartItem"
>
  <img
    class="sc-grXZZQ bVmYJu"
    src="https://via.placeholder.com/150"
  />
  <div
    class="sc-dkjKgF dJrBAg"
  >
    <div
      class="sc-hLclGa eunyYl"
    >
      <h4
        class="sc-eDnVMP jOpDHu"
      >
        Camiseta Ramones
      </h4>
      <button
        class="sc-jSwlEQ cQZDuO"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16.1378 21H7.85782C6.81082 21 5.94082 20.192 5.86282 19.147L4.96582 7H18.9998L18.1328 19.142C18.0578 20.189 17.1868 21 16.1378 21V21Z"
            fill-rule="evenodd"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M12 11V17"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M4 7H20"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M17 7L15.987 4.298C15.694 3.517 14.948 3 14.114 3H9.886C9.052 3 8.306 3.517 8.013 4.298L7 7"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M15.4298 11L14.9998 17"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M8.57016 11L9.00016 17"
            stroke="#DE3838"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </button>
    </div>
    <p
      class="sc-gKHVLF eaMKhs"
    >
      Aut dolor eum aspernatur maxime in suscipit perspiciatis aut iusto. Eum odio ducimus nam quos quia esse quibusdam totam iure. Voluptates est sit laboriosam. Et unde illum aliquid debitis quia fugit est nisi provident.
    </p>
    <div
      class="sc-iBAaJG blTQXt"
    >
      <select
        class="sc-pqitP kqyUUr"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="7"
        >
          7
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="9"
        >
          9
        </option>
        <option
          value="10"
        >
          10
        </option>
      </select>
      <span
        class="sc-ftLKQv cOBslt"
      >
        R$ 10,00
      </span>
    </div>
  </div>
</div>
`;
